baseplate/observers/metrics_tagged.py

```diff
@@ -0,0 +1,187 @@
+from baseplate import BaseplateObserver
+from baseplate import LocalSpan
+from baseplate import RequestContext
+from baseplate import Span
+from baseplate import SpanObserver
+from baseplate.lib import config
+from baseplate.lib import metrics
+from baseplate.observers.timeout import ServerTimeout
+
+
+logger = logging.getLogger(__name__)
+
+
+class Errors(Enum):
+    TIMED_OUT = 1
+    CONCURRENCY_LIMITED = 2
+    EXCEPTION = 3
+
+
+error_messages = {
+    Errors.TIMED_OUT: "timed_out",
+    Errors.CONCURRENCY_LIMITED: "concurrency_limited",
+    Errors.EXCEPTION: "internal_server_error",
+}
+
+
+class TaggedMetricsBaseplateObserver(BaseplateObserver):
+    """Metrics collecting observer.
+
+    This observer reports metrics to statsd in the dogstatsd format. It does three important things:
+
+    * it tracks the time taken in serving each request.
+    * it batches all metrics generated during a request into as few packets
+      as possible.
+    * it adds tags to the metric if they are in the config tag whitelist
+
+    The batch is accessible to your application during requests as the
+    ``metrics`` attribute on the :py:class:`~baseplate.RequestContext`.
+
+    :param client: The client where metrics will be sent.
+    :param cfg: the parsed application config with the tag whitelist
+
+    """
+
+    def __init__(self, client: metrics.Client, whitelist: List[str]):
+        self.client = client
+        self.whitelist = whitelist
+
+    @classmethod
+    def from_config_and_client(
+        cls, raw_config: config.RawConfig, client: metrics.Client
+    ) -> "TaggedMetricsBaseplateObserver":
+        cfg = config.parse_config(
+            raw_config,
+            {
+                "metrics": {
+                    "whitelist": config.Optional(
+                        config.TupleOf(config.String), default="endpoint,success,error"
+                    ),
+                }
+            },
+        )
+        return cls(client, whitelist=cfg.metrics.whitelist)
+
+    def on_server_span_created(self, context: RequestContext, server_span: Span) -> None:
+        batch = self.client.batch()
+        context.metrics = batch
+        observer = TaggedMetricsServerSpanObserver(batch, server_span, self.whitelist)
+        server_span.register(observer)
+
+
+class TaggedMetricsServerSpanObserver(SpanObserver):
+    def __init__(self, batch: metrics.Batch, server_span: Span, whitelist: list):
+        self.batch = batch
+        self.span = server_span
+        self.base_name = "baseplate.server"
+        self.whitelist = whitelist
+        self.tags = {}  # type: Dict[str, str]
+        self.timer = batch.timer(self.base_name)
+
+    def on_start(self) -> None:
+        self.span.set_tag("endpoint", self.span.name)
+        self.timer.start()
+
+    def on_incr_tag(self, key: str, delta: float) -> None:
+        self.batch.counter(key, self.tags).increment(delta, True)
+
+    def on_set_tag(self, key: str, value: Any) -> None:
```

Comment:
would it make sense to call `update_tags()` directly in here to get those tags onto the timer so we don't have to build up a parallel dictionary of stuff in this object?

Reply:
currently update_tags completely replaces its current tags with a new dict of tags, so callign it would lose any existing tags already on the timer. I'm open to changing it so that update_tags is smarter and adds if not already on the tags and replaces if it is though. Not sure of which is preferable

Reply:
ah yeah, if update tags were like:

```python
def update_tags(self, tags):
    self.tags.update(tags)
```

that should do the trick?

Reply:
ah, i'm sorry, i mislead you here. the update_tags as-is works for the timer but not the counters. i think we do have to .update() on the self.tags and like you originally had.  sorry :( (the rest is fine though)